File: Web/src/api/notes.ts
import { api } from "./client";

export interface NoteStockMention {
  code: string;
  name: string;
}

export interface NoteSummary {
  id: string;
  title: string;
  trade_date: string | null;
  tags: string[];
  linked_codes: string[];
  mentions?: NoteStockMention[];
  excerpt: string;
  created_at: string;
  updated_at: string;
}

export interface NoteDetail {
  id: string;
  title: string;
  content_html: string;
  trade_date: string | null;
  tags: string[];
  linked_codes: string[];
  created_at: string;
  updated_at: string;
}

export interface NoteListResponse {
  items: NoteSummary[];
  total: number;
  page: number;
  page_size: number;
}

export interface NoteCreatePayload {
  title: string;
  content_html: string;
  trade_date?: string | null;
  tags?: string[];
}

export interface NoteUpdatePayload {
  title?: string;
  content_html?: string;
  trade_date?: string | null;
  tags?: string[];
}

export async function listNotes(params?: {
  q?: string;
  tag?: string;
  code?: string;
  page?: number;
  page_size?: number;
}): Promise<NoteListResponse> {
  const { data } = await api.get<NoteListResponse>("/notes", { params });
  return data;
}

export async function listNoteTags(): Promise<string[]> {
  const { data } = await api.get<string[]>("/notes/tags");
  return data;
}

export async function getNote(id: string): Promise<NoteDetail> {
  const { data } = await api.get<NoteDetail>(`/notes/${id}`);
  return data;
}

export async function createNote(payload: NoteCreatePayload): Promise<NoteDetail> {
  const { data } = await api.post<NoteDetail>("/notes", payload);
  return data;
}

export async function updateNote(id: string, payload: NoteUpdatePayload): Promise<NoteDetail> {
  const { data } = await api.put<NoteDetail>(`/notes/${id}`, payload);
  return data;
}

export async function deleteNote(id: string): Promise<void> {
  await api.delete(`/notes/${id}`);
}

File: backend/app/api/notes.py
"""复盘笔记 API。"""
from __future__ import annotations

from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.orm import Session

from ..core.database import get_session
from ..schemas.note import (
    NoteCreateRequest,
    NoteDetail,
    NoteListResponse,
    NoteSummary,
    NoteUpdateRequest,
)
from ..services import note_service

router = APIRouter()


@router.get("/notes/tags", response_model=list[str])
def list_tags(session: Session = Depends(get_session)) -> list[str]:
    return note_service.list_all_tags(session)


@router.get("/notes", response_model=NoteListResponse)
def list_notes(
    q: str = Query(default="", max_length=100),
    tag: str = Query(default="", max_length=50),
    code: str = Query(default="", max_length=16),
    page: int = Query(default=1, ge=1),
    page_size: int = Query(default=20, ge=1, le=100),
    session: Session = Depends(get_session),
) -> NoteListResponse:
    items, total = note_service.list_notes(
        session, q=q, tag=tag, code=code, page=page, page_size=page_size
    )
    return NoteListResponse(
        items=[NoteSummary.model_validate(i) for i in items],
        total=total,
        page=page,
        page_size=page_size,
    )


@router.post("/notes", response_model=NoteDetail)
def create_note(
    req: NoteCreateRequest,
    session: Session = Depends(get_session),
) -> NoteDetail:
    note = note_service.create_note(
        session,
        title=req.title,
        content_html=req.content_html,
        trade_date=req.trade_date,
        tags=req.tags,
    )
    session.commit()
    return NoteDetail.model_validate(note)


@router.get("/notes/{note_id}", response_model=NoteDetail)
def get_note(
    note_id: str,
    session: Session = Depends(get_session),
) -> NoteDetail:
    note = note_service.get_note(session, note_id)
    if note is None:
        raise HTTPException(404, "note not found")
    return NoteDetail.model_validate(note)


@router.put("/notes/{note_id}", response_model=NoteDetail)
def update_note(
    note_id: str,
    req: NoteUpdateRequest,
    session: Session = Depends(get_session),
) -> NoteDetail:
    note = note_service.get_note(session, note_id)
    if note is None:
        raise HTTPException(404, "note not found")
    note_service.update_note(
        session,
        note,
        title=req.title,
        content_html=req.content_html,
        trade_date=req.trade_date,
        tags=req.tags,
    )
    session.commit()
    return NoteDetail.model_validate(note)


@router.delete("/notes/{note_id}")
def delete_note(
    note_id: str,
    session: Session = Depends(get_session),
) -> dict[str, str]:
    note = note_service.get_note(session, note_id)
    if note is None:
        raise HTTPException(404, "note not found")
    note_service.delete_note(session, note)
    session.commit()
    return {"status": "ok"}
